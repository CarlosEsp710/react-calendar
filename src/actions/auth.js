import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLoginAction = (email, password) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken("auth/login", "POST", {
      email,
      password,
    });

    const data = await res.json();

    if (data.ok) {
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        loginAction({
          uid: data.uid,
          email: data.email,
          name: data.name,
        })
      );
    } else {
      Swal.fire("Error", data.msg, "error");
    }
  };
};

const loginAction = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startRegisterAction = (name, email, password) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken("auth/register", "POST", {
      name,
      email,
      password,
    });

    const data = await res.json();

    if (data.ok) {
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        registerAction({
          uid: data.uid,
          email: data.email,
          name: data.name,
        })
      );
    } else {
      Swal.fire("Error", data.msg, "error");
    }
  };
};

const registerAction = (user) => ({
  type: types.authRegister,
  payload: user,
});

export const startCheckTokenAction = () => {
  return async (dispatch) => {
    const res = await fetchWithToken("auth/renew-token", "GET", {});
    const data = await res.json();

    if (data.ok) {
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        loginAction({
          uid: data.uid,
          email: data.email,
          name: data.name,
        })
      );
    } else {
      Swal.fire("Error", data.msg, "error");
      dispatch(checkTokenAction());
    }
  };
};

const checkTokenAction = () => ({
  type: types.authCheckToken,
});

export const startLogoutAction = () => {
  return (dispatch) => {
    localStorage.removeItem("token");
    localStorage.removeItem("token-init-date");

    dispatch(logoutAction());
  };
};

const logoutAction = () => ({
  type: types.authLogout,
});
