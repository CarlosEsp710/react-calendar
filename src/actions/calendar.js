import Swal from "sweetalert2";

import { fetchWithToken } from "../helpers/fetch";
import { loadEvents } from "../helpers/loadEvents";
import { types } from "../types/types";

export const startLoadingEventsAction = () => {
  return async (dispatch) => {
    const res = await fetchWithToken("events", "GET", {});
    const data = await res.json();

    if (data.ok) {
      const events = loadEvents(data.events);
      dispatch(loadEventsAction(events));
    } else {
      Swal.fire("Error", data.msg, "error");
    }
    try {
    } catch (error) {
      Swal.fire("Error", error.message, "error");
    }
  };
};

const loadEventsAction = (events) => ({
  type: types.calendarGetEvents,
  payload: events,
});

export const startAddEventAction = (event) => {
  return async (dispatch, getState) => {
    const { uid, name, email } = getState().auth;

    try {
      const res = await fetchWithToken("events", "POST", event);
      const data = await res.json();

      if (data.ok) {
        event.uid = data.event.uid;
        event.user = { uid, name, email };
        dispatch(addEventAction(event));
      } else {
        Swal.fire("Error", data.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
    }
  };
};

const addEventAction = (event) => ({
  type: types.calendarAddEvent,
  payload: event,
});

export const startUpdateEventAction = (event) => {
  return async (dispatch, getState) => {
    const { uid, name, email } = getState().auth;

    try {
      const res = await fetchWithToken(`events/${event.uid}`, "PUT", event);
      const data = await res.json();

      if (data.ok) {
        event.uid = data.event.uid;
        event.user = { uid, name, email };
        dispatch(updateEventAction(event));
      } else {
        Swal.fire("Error", data.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
    }
  };
};

const updateEventAction = (event) => ({
  type: types.calendarUpdateEvent,
  payload: event,
});

export const startDeleteEventAction = (event) => {
  return async (dispatch, getState) => {
    const { activeEvent } = getState().calendar;

    try {
      const res = await fetchWithToken(
        `events/${activeEvent.uid}`,
        "DELETE",
        event
      );
      const data = await res.json();

      if (data.ok) {
        dispatch(deleteEventAction(event));
      } else {
        Swal.fire("Error", data.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", error.message, "error");
    }
  };
};

const deleteEventAction = () => ({
  type: types.calendarDeleteEvent,
});

export const setActiveEventAction = (event) => ({
  type: types.calendarSetActiveEvent,
  payload: event,
});

export const clearActiveEventAction = () => ({
  type: types.calendarClearActiveEvent,
});

export const clearCalendarAction = () => ({
  type: types.calendarClear,
});
