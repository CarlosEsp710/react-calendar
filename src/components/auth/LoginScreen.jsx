import React from "react";

import { useDispatch } from "react-redux";
import Swal from "sweetalert2";

import { useForm } from "../../hooks/useForm";
import { startLoginAction, startRegisterAction } from "../../actions/auth";

import "../../assets/styles/login.css";

export const LoginScreen = () => {
  const dispatch = useDispatch();

  const [loginForm, handleLoginInputChange, resetLogin] = useForm({
    loginEmail: "carlos@gmail.com",
    loginPassword: "123456",
  });

  const [registerForm, handleRegisterInputChange, resetRegister] = useForm({
    registerName: "Carlos",
    registerEmail: "carlos@gmail.com",
    registerPassword: "123456",
    registerConfirmPassword: "123456",
  });

  const { loginEmail, loginPassword } = loginForm;
  const {
    registerName,
    registerEmail,
    registerPassword,
    registerConfirmPassword,
  } = registerForm;

  const handleLogin = (event) => {
    event.preventDefault();
    dispatch(startLoginAction(loginEmail, loginPassword));
    resetLogin();
  };

  const handleRegister = (event) => {
    event.preventDefault();

    if (registerPassword !== registerConfirmPassword) {
      return Swal.fire("Error", "Las contrase単as no coinciden", "error");
    }

    dispatch(
      startRegisterAction(registerName, registerEmail, registerPassword)
    );
    resetRegister();
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Ingreso</h3>
          <form onSubmit={handleLogin}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
                name="loginEmail"
                value={loginEmail}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contrase単a"
                name="loginPassword"
                value={loginPassword}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="registerName"
                value={registerName}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Correo"
                name="registerEmail"
                value={registerEmail}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contrase単a"
                name="registerPassword"
                value={registerPassword}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contrase単a"
                name="registerConfirmPassword"
                value={registerConfirmPassword}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Crear cuenta" />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
