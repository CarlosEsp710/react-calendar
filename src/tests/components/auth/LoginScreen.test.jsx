import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { LoginScreen } from "../../../components/auth/LoginScreen";
import { startLoginAction } from "../../../actions/auth";

jest.mock("../../../actions/auth", () => ({
  startLoginAction: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <LoginScreen />
  </Provider>
);

describe("test <LoginScreen /> component", () => {
  test("should render <LoginScreen /> component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call startLoginAction", () => {
    wrapper
      .find("form")
      .at(0)
      .simulate("submit", {
        preventDefault: () => {},
      });

    expect(startLoginAction).toHaveBeenCalled();
  });
});
