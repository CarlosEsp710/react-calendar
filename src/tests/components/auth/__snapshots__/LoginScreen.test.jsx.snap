// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <LoginScreen /> component should render <LoginScreen /> component 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value="carlos@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="loginPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Carlos"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="carlos@gmail.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerConfirmPassword"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value="123456"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
